Testing hz applications (#1762)
hazelcast/hz-docs: docs/modules/test/pages/testing-common.adoc

[source,java]
----
@RunWith(HazelcastParallelClassRunner.class)
public class MyClusterTest extends HazelcastTestSupport {
    // ...
}
----

Tests run with these runners may fail if some other test doesn't shutdown client or member instances correctly.

== Isolating Clusters by Name

Use `randomName()` to avoid interference when tests run in parallel.

[source,java]
----
String clusterName = HazelcastTestSupport.randomName();
Config config = new Config().setClusterName(clusterName);
HazelcastInstance[] members = factory.newInstances(config, 2);

ClientConfig clientConfig = new ClientConfig().setClusterName(clusterName);
HazelcastInstance client = factory.newHazelcastClient(clientConfig);
----

=== Cleaning Up Resources

Use `shutdownAll()` to clean up all Hazelcast instances at the end of each test or test class.

JUnit 4:
[source,java]
----
@After
public void tearDown() {
    client.shutdown();
    member1.shutdown();

    // or
    shutdownGracefullyNodeFactory();
}
----

JUnit 5:
[source,java]
----
@AfterAll
static void tearDownCluster() {
    client.shutdown();
    member1.shutdown();

    // or
    factory.shutdownAll();
}
----

> Shutting down a single instance can be used to mimic failure scenarios `ìnstance.shutdown()`.

Reviewer: Missing words, or new line with code sample?

Author: I'll fix the content in a new commit

= Common test utilities
:description: This section describes Hazelcast's common test utilities across caching and streaming applications.

{description}

== Repeat tests

> Use `@Repeat(n)` (`com.hazelcast.test.annotation.Repeat`) to repeat a test multiple times. This is useful to execute a test multiple time to trigger flaky behaviour. The test passes _only_ if all repeated execution succeed.

Reviewer 2: ```suggestion
Use `@Repeat(n)` (`com.hazelcast.test.annotation.Repeat`) to repeat a test multiple times. This is useful to execute a test multiple times to trigger flaky behavior. The test passes _only_ if all repeated executions succeed.
```

[source,java]
----
@Repeat(5)
@Test
public void testFlakyBehaviour() {
    // ...
}
----

== Test runners

Hazelcast provides custom runners:

- `HazelcastSerialClassRunner`: runs tests sequentially.
- `HazelcastParallelClassRunner`: runs tests in parallel.

[source,java]
----
@RunWith(HazelcastParallelClassRunner.class)
public class MyClusterTest extends HazelcastTestSupport {
    // ...
}
----

Tests using these runners may fail if some other test doesn't shutdown client or member instances correctly.

== Isolate clusters by name

Use `randomName()` to avoid interference when tests run in parallel.

[source,java]
----
String clusterName = HazelcastTestSupport.randomName();
Config config = new Config().setClusterName(clusterName);
HazelcastInstance[] members = factory.newInstances(config, 2);

ClientConfig clientConfig = new ClientConfig().setClusterName(clusterName);
HazelcastInstance client = factory.newHazelcastClient(clientConfig);
----

=== Clean up resources

Use `shutdownAll()` to clean up all Hazelcast instances at the end of each test or test class.

JUnit 4:
[source,java]
----
@After
public void tearDown() {
    client.shutdown();
    member1.shutdown();

    // or

>     shutdownGracefullyNodeFactory();

Author: fixed

}
----

JUnit 5:
[source,java]
----
@AfterAll
static void tearDownCluster() {
    client.shutdown();
    member1.shutdown();

    // or
    factory.shutdownAll();
}
----

> Shutting down a single instance (via `ìnstance.shutdown()`) can be used to mimic failure scenarios.

Reviewer 3: ```suggestion
Shutting down a single instance via `ìnstance.shutdown()` will trigger a graceful shutdown. The method `instance.getLifecycleService().terminate()` can be used to trigger an ungraceful termination, which can be used to mimic failure scenarios.

Author: cool thanks! fixed

[source,java]
----
import com.hazelcast.config.Config;
import com.hazelcast.core.HazelcastInstance;
import com.hazelcast.map.MapStore;
import com.hazelcast.test.TestHazelcastFactory;
import org.junit.jupiter.api.*;

import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

class MyMapStoreTest {

    private static TestHazelcastFactory factory;
    private static HazelcastInstance hz;

    @BeforeAll
    static void setup() {
        factory = new TestHazelcastFactory();

        // Create a mock MapStore that throws on load
        @SuppressWarnings("unchecked")
        MapStore<String, String> mockMapStore = mock(MapStore.class);
        when(mockMapStore.load("fail")).thenThrow(new RuntimeException("Simulated failure"));
        when(mockMapStore.load("key1")).thenReturn("value1");

        // Configure Hazelcast to use the mock MapStore
        Config config = new Config();

>         config.setClusterName("mock-mapstore-test");

Reviewer 3: Nit: This slightly goes against our best practice given of using `randomName()`

        config.getMapConfig("testMap")
              .getMapStoreConfig()
              .setEnabled(true)
              .setImplementation(mockMapStore);

        hz = factory.newHazelcastInstance(config);
    }

    @AfterAll
    static void teardown() {

>         if (hz != null) {
>             hz.shutdown();
>         }
>         factory.shutdownAll();

Reviewer 3: ```suggestion
        if (factory != null) {
            factory.shutdownAll();
        }
```
A factory shutdown is sufficient, but it can also be null (could be initialized statically)

Author: done

    }

    @Test
    void testSuccessfulLoadFromMock() {
        var map = hz.getMap<String, String>("testMap");

        // This triggers MapStore.load("key1")
        String result = map.get("key1");
        assertEquals("value1", result);
    }

    @Test
    void testLoadFailureHandled() {
        var map = hz.getMap<String, String>("testMap");

        RuntimeException ex = assertThrows(RuntimeException.class, () -> {
            map.get("fail"); // triggers MapStore.load("fail")
        });

>         assertEquals("Simulated failure", ex.getMessage());

Reviewer 3: Not sure if we care to include it, but we have AssertJ assertions in Hazelcast test scope too, and that provides a really nice utility so you can do this here instead:
```
        assertThatThrownBy(() ->  map.get("fail"))
                .isInstanceOf(RuntimeException.class)
                .hasMessage("Simulated failure");
```
